tools/plagiarism-checker/src/index.ts

```diff
@@ -0,0 +1,93 @@
+import { Hono } from "hono"
+
+interface SearchResult {
+  title?: string | null
+  link?: string | null
+}
+
+interface SearchResponse {
+  items: Array<SearchResult>
+}
+
+interface PlagiarismResults {
+  sentence: string
+  matches: Array<SearchResult>
+}
+
+type Env = {
+  GOOGLE_API_KEY: string
+  GOOGLE_SEARCH_ENGINE_CX: string
+}
+
+const app = new Hono<{ Bindings: Env }>()
+
+app.use("*", async (c, next) => {
+  const apiKey = c.env.GOOGLE_API_KEY
+  if (!apiKey) {
+    return c.text("API key not found", 400)
+  }
+
+  const searchEngine = c.env.GOOGLE_SEARCH_ENGINE_CX
+  if (!searchEngine) {
+    return c.text("Search engine not specified", 400)
+  }
+
+  return next()
+})
+
+app.post("/", async (c) => {
+  const { text } = await c.req.json()
+
+  if (typeof text !== "string") {
+    return c.json({ error: "Invalid JSON format" }, 400)
+  }
+
+  const sentences = text
+    .split(".")
+    // if it's less than 10 characters - we don't treat it as a sentence and skip it
+    .filter((sentence) => sentence.trim().length > 10)
+
+  const plagiarismResults: PlagiarismResults[] = []
+
+  await Promise.all(
+    sentences.map(async (sentence) => {
+      const query = encodeURIComponent(sentence.trim())
+      const url = `https://www.googleapis.com/customsearch/v1?exactTerms=${query}&key=${c.env.GOOGLE_API_KEY}&cx=${c.env.GOOGLE_SEARCH_ENGINE_CX}&num=1`
+      const response = await fetch(url)
+      const searchResults: SearchResponse = await response.json()
+
+      if (!searchResults.items) {
+        return
+      }
+      plagiarismResults.push({
+        sentence,
+        matches: searchResults.items.map((item: SearchResult) => ({
+          title: item.title,
+          link: item.link
+        }))
+      })
+    })
+  )
```

Comment:
Modifying a shared resource like `plagiarismResults` within multiple asynchronous operations can sometimes lead to race conditions.

Reply:
I changed it, now we are not pushing to an external array, but returning from Promise.all directly to a variable. Then we are filtering out null values from the array.